### DISCUSSION TITLE
receiving and transmitting a file using FSK

### DISCUSSION THREAD

**Discussion by LeonidLd** [duggabe/gr-control#13] [Q&A] [answered]
Good afternoon We are trying to transfer a text file using FSK modulation. At the reception it is not possible to pass the "Access code" block
We transmit the signal at a frequency of 433.2 MHz, sampling frequency 8e6. Can you tell me what the problem might be?
![RX_cx](https://github.com/duggabe/gr-control/assets/122673719/da547852-8a7e-4d3f-9eec-9c52a6d5404b)
![RX](https://github.com/duggabe/gr-control/assets/122673719/b74eb857-5a23-43d1-9835-db2244b36a26)
![TX_cx](https://github.com/duggabe/gr-control/assets/122673719/52ebaeda-a458-413c-88b0-d118aecc9f68)

**Comment by duggabe** [comment]
Do you really need such a high sample rate for the Hack RF?

Using a much lower sample rate which is an integer multiple of 48kHz will reduce the Decimation factor for the Rational Resampler. The current factor of 3/500 creates too many taps for one block.

On the transmit side, you can use a Repeat block instead of the Rational Resampler for a more efficient block.

For my USRP B200mini, I use a sample rate of 768kHz, so the total Interpolation/Decimation factor is only 16.

Also, I recommend testing with the ZMQ blocks in place of the Soapy blocks just to make sure everything else is working. When you **are** using the Soapy blocks, the Throttle blocks should not be needed.

**Reply by LeonidLd** [reply]
Good afternoon thanks for your reply! We tried to test the circuit with ZMQ blocks (Figure 1). The figure shows the output from HackRF, after Filter, and the output after the Quadrature Demod block. Everything works well.
We also placed the Repeat block in place of the Rational Resampler block and changed the sampling frequency to 1e6 - this is the minimum SR frequency in HackRF (The results are shown in Figure 2).
Our results differ; the waterfall representation of the signal using HackRF looks like this (Figure 3.) Our signal seems to shift to the left plane, this can also be seen in the spectrogram. Can you tell me what could be the problem here?
![ZQM](https://github.com/duggabe/gr-control/assets/122673719/9a475fd3-1175-4178-b18c-a65832fc36e7) - Рисунок.1.
![repeat](https://github.com/duggabe/gr-control/assets/122673719/ab100fe4-0d04-44dd-9588-ebc9e3241eea) - Рисунок .2.
![рис3](https://github.com/duggabe/gr-control/assets/122673719/7ab7e016-0727-48f6-b0c7-8103cc061a82) - Рисунок 3.

**Comment by duggabe** [comment]
The frequency shift is because this is AFSK with a center frequency of 1kHz.

I looked up sample rates for the HackRF at https://hackrf.readthedocs.io/en/latest/sampling_rate.html and it looks like you DO need to use 8MHz.

Here is what I suggest:

Study https://wiki.gnuradio.org/index.php?title=Sample_Rate_Tutorial#Source_hardware_example and the following two sections to learn about sample rates in GNU Radio.

You have an old version of the packet and FSK code. So:
- delete your copy of gr-control
- follow the instructions in https://wiki.gnuradio.org/index.php?title=Simulation_example:_FSK#File_transfer_using_Packet_and_AFSK **exactly**.
At the end of that tutorial, you should have a file named 'output.png' which is our GNU Radio icon.

Now let's change some things to work with your HackRF:
For the transmitter:

    change 'samp_rate' to 80000
    change 'baud' to 1000
    change the VCO (complex) Amplitude to 0.9
    replace the Throttle with a Repeat, Interpolation = 100
    replace the ZMQ PUB Sink with your Soapy HackRF Sink (Sample Rate = 8MHz)
    click the "Generate the Flowgraph" icon

For the receiver:

    change 'samp_rate' to 80000
    change 'baud' to 1000
    change 'decim' to 4
    replace the ZMQ SUB Source with your Soapy HackRF Source (Sample Rate = 8MHz)
    add a Rational Resampler
        Interpolation = 1
        Decimation = 100
    click the "Generate the Flowgraph" icon
    delete ~/gr-control/Receivers/output.* files

Repeat the process starting with https://wiki.gnuradio.org/index.php?title=Simulation_example:_FSK#Receiving_the_File
At the end of that tutorial, you should have a file named 'output.png' which is our GNU Radio icon.

You can transfer any file by changing the filename in the transmit command line, e.g.:

    python3 pkt_fsk_xmt.py --InFile="../GRCon21_presentation.pdf"

Of course, in the 'strip_preamble' command you should change the output file name extension to .pdf 

Either relative or absolute file paths work, but using a tilde (~) does not.

**Reply by LeonidLd** [reply]
Good afternoon thanks for the answer! We did everything as you said, changed the values ​​on reception and transmission, and also read your links. In Figure 1 We used Soapy HackRF and transferred between two HackRFs (HackRF parameters in Figure 3.) In Figure 2 we used transfer to ourselves and using ZMQ blocks. We are confused by the output signal after the Quadrature Demod block, as if it cannot determine when it turns out to be “0”. Can you tell me what the problem might be?



![photo_2024-07-10_12-01-26](https://github.com/duggabe/gr-control/assets/122673719/a16ceee9-1264-48dd-9216-a232f512a348) - Рисунок.1.
![image](https://github.com/duggabe/gr-control/assets/122673719/5285df27-d302-4bed-bf3b-77f0485b2160) - Рисунок.2.
![image](https://github.com/duggabe/gr-control/assets/122673719/ba62fcab-4ef9-4e12-b72b-6c5e37d91d3e) - Рисунок.3.1
![image](https://github.com/duggabe/gr-control/assets/122673719/9bbd311b-7046-4e14-aea8-d70e0ad89167) - Рисунок.3.2

**Comment by duggabe** [comment, marked as answer]
I think you are very close :)
With AFSK, the tuning of the transmit and receive frequencies is critical. I suggest that you add a QT GUI Range block to the receiver with:
- Id = rx_freq
- Label = Receive freq
- Type = float
- Default Value = 429.22e6
- Start = 429.21e6
- Stop = 429.23e6
- Step = 100
- Widget = Counter

and then change the Soapy HackRF Source Center Freq to 'rx_freq'
click the "Generate the Flowgraph" icon

While the data is transferring you can adjust the Receive freq in small steps to center the output of the Quadrature Demod block. At that point you should get output from your Correlate Access Code - Tag Stream block.

Then, remembering the frequency which worked best, stop the transmitter and receiver, restart the receiver, set the correct frequency with the QT GUI Range block, and start the transmitter.

**Reply by LeonidLd** [reply]
Thanks a lot! You helped us a lot!
